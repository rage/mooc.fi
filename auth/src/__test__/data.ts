import { NextFunction, Request, Response } from "express"

export const testProfile = {
  edu_person_affiliation: "",
  edu_person_principal_name: "test",
  schac_home_organization: "helsinki.fi",
  organizational_unit: "organizational_unit",
  schac_personal_unique_code:
    "urn:schac:personalUniqueCode:int:studentID:helsinki.fi:121345678",
  display_name: "test",
  given_name: "test",
  mail: "test@mail.com",
}

export const req = {
  body: {},
  query: {
    RelayState: JSON.stringify({
      language: "fi",
      provider: "hy",
      action: "sign-in",
    }),
  },
  params: {},
  login: jest.fn().mockImplementation((_: any, fn: Function) => fn(null)),
  logout: jest.fn(),
} as unknown as Request
export const res = {
  redirect: jest.fn().mockReturnThis(),
  locals: {},
  setMOOCCookies: jest.fn().mockReturnThis(),
} as unknown as Response
export const next = jest.fn() as NextFunction

export const okTokenResponse = {
  data: {
    status: 200,
    success: true,
    tmc_token: "tmc_token",
    access_token: "access_token",
    admin: false,
  },
}
