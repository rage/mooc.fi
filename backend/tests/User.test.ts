import { gql } from "graphql-request"
import { getTestContext } from "./__helpers"
import { adminUserDetails, normalUser, normalUserDetails } from "./data"
import nock from "nock"

const addUserMutation = gql`
  mutation AddUser($user: UserArg!) {
    addUser(user: $user) {
      id
      email
      first_name
      last_name
      username
      research_consent
      upstream_id
      administrator
    }
  }
`

const updateReseachConsentMutation = gql`
  mutation UpdateResearchConsent($value: Boolean!) {
    updateResearchConsent(value: $value) {
      id
    }
  }
`

const ctx = getTestContext()

beforeAll(() => {
  nock("https://tmc.mooc.fi")
    .get("/api/v8/users/current?show_user_fields=1&extra_fields=1")
    .reply(function () {
      const auth = this.req.headers.authorization

      switch (auth) {
        case "Bearer normal":
          return [200, normalUserDetails]
        case "Bearer admin":
          return [200, adminUserDetails]
      }
    })
})

afterAll(() => nock.cleanAll())

describe("user queries", () => {
  describe("currentUser", () => {
    beforeEach(async () => {
      await ctx.prisma.user.create({
        data: normalUser,
      })
    })

    it("shows current user when logged in", async () => {
      ctx!.client.setHeader("Authorization", "Bearer normal")

      const res = await ctx.client.request(`
      query {
        currentUser {
          id
          administrator
          email
          first_name
          last_name
          username
          upstream_id
        }
      }
    `)

      expect(res).toMatchInlineSnapshot(
        {
          currentUser: {
            id: expect.any(String),
            administrator: false,
            email: "e@mail.com",
            first_name: "first",
            last_name: "last",
            username: "user",
            upstream_id: 1,
          },
        },
        `
              Object {
                "currentUser": Object {
                  "administrator": false,
                  "email": "e@mail.com",
                  "first_name": "first",
                  "id": Any<String>,
                  "last_name": "last",
                  "upstream_id": 1,
                  "username": "user",
                },
              }
          `,
      )

      jest.clearAllMocks()
    })

    it("shows null when not logged in", async () => {
      ctx!.client.setHeader("Authorization", "")

      const res = await ctx.client.request(`
      query {
        currentUser {
          id
          administrator
          email
          first_name
          last_name
          username
          upstream_id
        }
      }
    `)

      expect(res).toMatchInlineSnapshot(
        {
          currentUser: null,
        },
        `
        Object {
          "currentUser": null,
        }
      `,
      )
    })
  })
})

describe("user mutations", () => {
  beforeEach(async () => {
    await ctx.prisma.user.deleteMany({ where: {} })
  })

  describe("addUser", () => {
    beforeAll(() => ctx!.client.setHeader("Authorization", ""))

    it("creates user correctly", async () => {
      const res = await ctx.client.request(addUserMutation, {
        user: {
          email: "e@mail.com",
          first_name: "first",
          last_name: "last",
          username: "username",
          research_consent: false,
          upstream_id: 1,
        },
      })

      expect(res).toMatchSnapshot({
        addUser: {
          id: expect.any(String),
        },
      })
    })

    it("won't create user with same id", async () => {
      await ctx.prisma.user.create({
        data: normalUser,
      })

      await expect(async () => {
        await ctx.client.request(addUserMutation, {
          user: {
            email: "e@mail.com",
            first_name: "first",
            last_name: "last",
            username: "username",
            research_consent: false,
            upstream_id: 1,
          },
        })
      }).rejects.toThrow()
    })
  })

  describe.skip("updateResearchConsent", () => {
    beforeEach(async () => {
      await ctx!.prisma.user.create({
        data: normalUser,
      })
    })

    afterEach(async () => {
      await ctx!.prisma.user.delete({ where: { upstream_id: 1 } })
      ctx.user = undefined
    })

    it("updates correctly", async () => {
      ctx!.client.setHeader("Authorization", "Bearer normal")

      const res = await ctx!.client.request(updateReseachConsentMutation, {
        value: true,
      })

      expect(res.updateResearchConsent).toMatchInlineSnapshot(
        { id: expect.any(String) },
        `
        Object {
          "id": Any<String>,
        }
      `,
      )
      const updatedConsent = await ctx!.prisma.user.findFirst({
        where: { upstream_id: 1 },
        select: { research_consent: true },
      })

      expect(updatedConsent).toMatchObject({ research_consent: true })
    })

    it("won't update research consent without auth", async () => {
      ctx!.client.setHeader("Authorization", "")

      try {
        await ctx!.client.request(updateReseachConsentMutation, { value: true })
        fail()
      } catch {}
    })
  })
})
