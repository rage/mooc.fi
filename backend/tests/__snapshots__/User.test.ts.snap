// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User mutations addUser creates user correctly 1`] = `
Object {
  "addUser": Object {
    "administrator": false,
    "email": "e@mail.com",
    "first_name": "first",
    "id": Any<String>,
    "last_name": "last",
    "research_consent": false,
    "upstream_id": 1,
    "username": "username",
  },
}
`;

exports[`User queries user_course_summary returns courses with completions 1`] = `
Object {
  "currentUser": Object {
    "id": "20000000-0000-0000-0000-000000000102",
    "user_course_summary": Array [
      Object {
        "course": Object {
          "id": "00000000-0000-0000-0000-000000000002",
          "name": "course1",
        },
        "course_id": "00000000-0000-0000-0000-000000000002",
        "exercise_completions": Array [
          Object {
            "exercise": Object {
              "id": "50000000-0000-0000-0000-000000000003",
              "max_points": null,
              "name": "exercise 3",
            },
            "exercise_completion_required_actions": Array [
              Object {
                "id": "66666666-0000-0000-0000-000000000003",
                "value": "TOO_MANY_DUCKS",
              },
            ],
            "id": "60000000-0000-0000-0000-000000000003",
            "n_points": 4,
            "timestamp": "1900-01-01T08:00:00.000Z",
          },
        ],
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      Object {
        "course": Object {
          "id": "00000000-0000-0000-0000-000000000001",
          "name": "course2",
        },
        "course_id": "00000000-0000-0000-0000-000000000001",
        "exercise_completions": Array [
          Object {
            "exercise": Object {
              "id": "50000000-0000-0000-0000-000000000001",
              "max_points": null,
              "name": "exercise 1",
            },
            "exercise_completion_required_actions": Array [],
            "id": "60000000-0000-0000-0000-000000000001",
            "n_points": 6,
            "timestamp": "1900-01-01T08:00:00.000Z",
          },
          Object {
            "exercise": Object {
              "id": "50000000-0000-0000-0000-000000000002",
              "max_points": null,
              "name": "exercise 2",
            },
            "exercise_completion_required_actions": Array [],
            "id": "60000000-0000-0000-0000-000000000002",
            "n_points": 0,
            "timestamp": "1900-01-01T08:00:00.000Z",
          },
        ],
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    ],
  },
}
`;
